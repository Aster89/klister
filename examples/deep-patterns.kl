#lang kernel

(datatype (Nat)
  (zero)
  (add1 (Nat)))

(define half
  (flet (half (n)
          (case n
            ((zero) (zero))
            ((add1 (zero)) (zero))
            ((add1 (add1 k)) (add1 (half k)))))
    half))

(example (half (add1 (add1 (add1 (add1 (zero)))))))
(example (half (add1 (add1 (add1 (add1 (add1 (zero))))))))

(define half*
  (flet (half* (n)
          (case n
            ((add1 (add1 k)) (add1 (half* k)))
            (_ (zero))))
    half*))

(example (half* (add1 (add1 (add1 (add1 (zero)))))))
(example (half* (add1 (add1 (add1 (add1 (add1 (zero))))))))


(define four?
  (lambda (x)
    (case x
      ((add1 (add1 (add1 (add1 (zero))))) (true))
      (_ (false)))))

(example (four? (add1 (add1 (add1 (add1 (zero)))))))
(example (four? (add1 (add1 (add1 (zero))))))
(example (four? (add1 (add1 (add1 (add1 (add1 (zero))))))))


(datatype (Pair A B)
  (pair A B))

(define reassoc
  (lambda (x)
    (case x
      ((pair (pair x y) z) (pair x (pair y z))))))

(example (reassoc (pair (pair 0 (zero)) "zero")))

