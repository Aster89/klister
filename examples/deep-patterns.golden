(add1 (add1 (zero))) : Nat
(add1 (add1 (zero))) : Nat
(add1 (add1 (zero))) : Nat
(add1 (add1 (zero))) : Nat
(true) : Bool
(false) : Bool
(false) : Bool
(pair 0 (pair (zero) "zero")) : (Pair Integer (Pair Nat String))
